server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistFileSourcesStep.java

```diff
@@ -0,0 +1,226 @@
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.step;
+
+import org.apache.commons.codec.binary.Hex;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.io.Charsets;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.LineIterator;
+import org.apache.commons.lang.StringUtils;
+import org.apache.ibatis.session.ResultContext;
+import org.apache.ibatis.session.ResultHandler;
+import org.sonar.api.resources.Qualifiers;
+import org.sonar.api.utils.System2;
+import org.sonar.batch.protocol.Constants;
+import org.sonar.batch.protocol.output.BatchReport;
+import org.sonar.batch.protocol.output.BatchReportReader;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.persistence.MyBatis;
+import org.sonar.core.source.db.FileSourceDto;
+import org.sonar.server.computation.ComputationContext;
+import org.sonar.server.computation.source.ReportIterator;
+import org.sonar.server.computation.source.StreamLine;
+import org.sonar.server.computation.source.StreamLineCoverage;
+import org.sonar.server.computation.source.StreamLineScm;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.source.db.FileSourceDb;
+
+import java.io.File;
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Charsets.UTF_8;
+import static com.google.common.collect.Lists.newArrayList;
+
+public class PersistFileSourcesStep implements ComputationStep {
+
+  private final DbClient dbClient;
+  private final System2 system2;
+
+  public PersistFileSourcesStep(DbClient dbClient, System2 system2) {
+    this.dbClient = dbClient;
+    this.system2 = system2;
+  }
+
+  @Override
+  public String[] supportedProjectQualifiers() {
+    return new String[] {Qualifiers.PROJECT};
+  }
+
+  @Override
+  public void execute(ComputationContext context) {
+    int rootComponentRef = context.getReportMetadata().getRootComponentRef();
+    // Don't use batch insert for file_sources since keeping all data in memory can produce OOM for big files
+    DbSession session = dbClient.openSession(false);
+    try {
+      final Map<String, FileSourceDto> previousFileSourcesByUuid = new HashMap<>();
+      session.select("org.sonar.core.source.db.FileSourceMapper.selectHashesForProject", context.getProject().uuid(), new ResultHandler() {
```

Comment:
No, we do not have all information in the index (missing src hash), and it has nothing to do in it. Moreover, as the goal is to update source file, it seems more safe to use the db as a referential.